python/cudf/cudf/testing/testing.py

```diff
@@ -201,31 +206,50 @@ def assert_column_equal(
     ):
         left = left.astype(left.categories.dtype)
         right = right.astype(right.categories.dtype)
-
     columns_equal = False
-    try:
-        columns_equal = (
-            (
-                cp.all(left.isnull().values == right.isnull().values)
-                and cp.allclose(
-                    left[left.isnull().unary_operator("not")].values,
-                    right[right.isnull().unary_operator("not")].values,
+    if left.size == right.size == 0:
+        columns_equal = True
+    elif not (
+        (is_string_dtype(left) and is_numeric_dtype(right))
+        or (is_numeric_dtype(left) and is_string_dtype(right))
+    ):
```

Comment:
Yeah this seems odd. Are we just generally looking to avoid checking this for mismatched dtypes? And doesn't cupy not support strings anyway, so if left and right are both strings we'll go into this block and then fail (the exception handling only deals with categoricals)?

Reply:
If one column is string and the other is numeric, we can assume the columns are not equal. (`1 != '1')`. These lines check to make sure exactly one of the columns is string and the other is numeric, in which case we avoid the entire try/except block and therefore avoid any opportunity for `columns_equal` to be set to `True`.  We should end up on line 243 from there.

Reply:
Is this only an issue for string types, or do we need to worry about other types as well? I see categoricals are handled above, what about list/struct dtypes?

Reply:
I looked into what happens here and it would appear that when `list` or `struct` are involved we fall into the `left.equals(right)` check at the very end and end up with a `TypeError`, except in the case that we're comparing `list` to `list` or `struct` to `struct`. That should probably not happen.

I see the point I think you are making though: there's a certain set of dtypes (beyond just string) where, even if `check_dtype=False`, we know up front that it's a 100% mismatch between non-null elements simply because a struct can't compare as equal to a "not struct". I think these dtypes are:

- String
- List
- Struct
- Interval
- Decimal

I suppose the only edge case here would be that we should arguably return `True` even for that subset of dtypes if the columns are fully null. 

If this seems like the correct logic I am happy to go back and wire it up as such here.

Reply:
Yep that all sounds good. For the fully null case, I think whether or not we return True should be determined by `check_dtypes`, but if someone has a different opinion I'm open to a different result.